pkgtools/digest/files/digest.1
.\" $NetBSD$
.\"
.\"
.\" Copyright (c) 2001 Alistair G. Crooks.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by Alistair G. Crooks.
.\" 4. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior written
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS
.\" OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
.\" DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
.\" GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\"
.Dd February 28, 2001
.Dt DIGEST 1
.Os NetBSD
.Sh NAME
.Nm digest
.Nd calculate message digests 
.Sh SYNOPSIS
.Nm ""
.Ar algorithm
.Op file...
.Sh DESCRIPTION
The
.Nm
utility calculates message digests of files or,
if no file is specified, standard input.
The list of possible algorithms is:
.Bl -tag -width Ds
.It md5
the
.Xr md5 3
algorithm will be used.
.It sha1
the
.Xr sha1 3
algorithm will be used.
.It rmd160
the
.Xr rmd160 3
algorithm will be used.
.El
.Pp
The
.Nm
utility is a simple wrapper for the various different
algorithm implementations, which are located in the standard
C library, and was designed to be scalable as new message digest
algorithms are developed.
.Pp
The
.Nm
utility exits 0 on success, and >0 if an error occurs.
.Sh SEE ALSO
.Xr cksum 1 ,
.Xr md5 3 ,
.Xr rmd160 3 ,
.Xr sha1 3
.Sh HISTORY
The
.Nm
utility first appeared in
.Nx 1.6 .
.Sh AUTHOR
The
.Nm
utility was written by Alistair G. Crooks (agc@netbsd.org).


doc/NEWS
# $NetBSD: NEWS,v 1.1 2008/01/18 05:47:30 tnn Exp $

PKGSRC NEWS
===========

  This file contains information about important infrastructural changes in
  pkgsrc. The intended readers are the pkgsrc developers as well as anyone
  tracking pkgsrc-current.


OpenSSL update [tnn 20080118]
	OpenSSL in pkgsrc switched from the 0.9.7 to the 0.9.8 branch. To
	accomodate for this, 596 packages dependent on openssl had their
	revisions bumped.

pkgsrc-2007Q4 released [agc 20080105]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2008/01/05/0000.html

Simplifying meta-packages [rillig 20071219]
	The variable META_PACKAGE was introduced. This simplifies creation
	and maintenance of packages that do not install files.

Replace a package with different name [gdt 20071107]
	The variable OLDNAME was introduced for "make replace" to
	replace a package with a name other than the new one, via
	"make replace OLDNAME=oldpkgname".  Read the announcement at:
	http://mail-index.netbsd.org/tech-pkg/2007/11/07/0002.html
	
pkgsrc-2007Q3 released [agc 20071015]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/10/15/0000.html

Features framework [jlam 20070907]
	A "system features" framework has been added. It will be used to
	automatically supply missing "basic" headers and libraries for old
	systems. This is supported through the USE_FEATURES variable.

X.org 6.9 removed [joerg 20070817]
	monolithic X.org 6.9 was removed from pkgsrc. It was obsoleted by the
	modular X.org (7.x) packages. Transitioning users should remove all old
	X.org packages and set X11_TYPE=modular in mk.conf.

Emulator framework [jlam 20070729]
	A new emulator framework has been added. This will handle all
	binary-only packages that require emulation on the native OS.
	Please see pkgsrc/mk/emulator/README.

audit-packages merged with pkg_install [adrianp 20070714]
	The utility for auditing installed packages for security problems,
	audit-packages, has been rewritten in C. It is now part of the
	pkgtools/pkg_install package.
	
pkgsrc-2007Q2 released [agc 20070701]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/07/01/0000.html

New framework for bulk builds [joerg 20070619]
	pbulk, the new automated pkgsrc bulk build framework was imported.

pkgsrc-2007Q1 released [agc 20070419]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/04/19/0000.html

pkgsrc-2006Q4 released [agc 20070104]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/01/04/0000.html

modular X.org imported [joerg 20061114]
	The modular X.org packages were imported.

pkgsrc-2006Q3 released [agc 20061001]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2006/10/01/0000.html


pkgtools/digest/files/digest.c
/*	$NetBSD$ */

/*
 * Copyright (c) 2001 Alistair G. Crooks.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by Alistair G. Crooks.
 * 4. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS
 * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
#include <sys/cdefs.h>

#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 2001 \
	        The NetBSD Foundation, Inc.  All rights reserved.");
__RCSID("$NetBSD$");
#endif

#include <sys/types.h>

#include <err.h>
#include <locale.h>
#include <md5.h>
#include <rmd160.h>
#include <sha1.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

/* perform an md5 digest, and print the results if successful */
static int
md5_digest_file(char *fn)
{
	MD5_CTX	m;
	char	in[BUFSIZ * 20];
	char	digest[33];
	int	cc;

	if (fn == NULL) {
		MD5Init(&m);
		while ((cc = read(STDIN_FILENO, in, sizeof(in))) > 0) {
			MD5Update(&m, (u_char *)in, (unsigned) cc);
		}
		(void) printf("%s\n", MD5End(&m, digest));
	} else {
		if (MD5File(fn, digest) == NULL) {
			return 0;
		}
		(void) printf("MD5 (%s) = %s\n", fn, digest);
	}
	return 1;
}

/* perform an sha1 digest, and print the results if successful */
static int
sha1_digest_file(char *fn)
{
	SHA1_CTX	sha;
	char		in[BUFSIZ * 20];
	char		digest[41];
	int		cc;

	if (fn == NULL) {
		SHA1Init(&sha);
		while ((cc = read(STDIN_FILENO, in, sizeof(in))) > 0) {
			SHA1Update(&sha, (u_char *)in, (unsigned) cc);
		}
		(void) printf("%s\n", SHA1End(&sha, digest));
	} else {
		if (SHA1File(fn, digest) == NULL) {
			return 0;
		}
		(void) printf("SHA1 (%s) = %s\n", fn, digest);
	}
	return 1;
}

/* perform an ripemd160 digest, and print the results if successful */
static int
rmd160_digest_file(char *fn)
{
	RMD160_CTX	rmd;
	char		in[BUFSIZ * 20];
	char		digest[41];
	int		cc;

	if (fn == NULL) {
		RMD160Init(&rmd);
		while ((cc = read(STDIN_FILENO, in, sizeof(in))) > 0) {
			RMD160Update(&rmd, (u_char *)in, (unsigned) cc);
		}
		(void) printf("%s\n", RMD160End(&rmd, digest));
	} else {
		if (RMD160File(fn, digest) == NULL) {
			return 0;
		}
		(void) printf("RMD160 (%s) = %s\n", fn, digest);
	}
	return 1;
}

/* this struct defines a message digest algorithm */
typedef struct alg_t {
	const char     *name;			/* algorithm name */
	int		(*func)(char *);	/* function to call */
} alg_t;

/* list of supported message digest algorithms */
static alg_t algorithms[] = {
	{ "md5",	md5_digest_file		},
	{ "rmd160",	rmd160_digest_file	},
	{ "sha1",	sha1_digest_file	},
	{ NULL	}
};

/* find an algorithm, given a name */
static alg_t *
find_algorithm(const char *a)
{
	alg_t	*alg;

	for (alg = algorithms ; alg->name && strcasecmp(alg->name, a) != 0 ; alg++) {
	}
	return (alg->name) ? alg : NULL;
}

int
main(int argc, char **argv)
{
	alg_t  *alg;
	int	rval;
	int	i;

	(void) setlocale(LC_ALL, "");
	while ((i = getopt(argc, argv, "V")) != -1) {
		switch(i) {
		case 'V':
			printf("%d\n", VERSION);
			return EXIT_SUCCESS;
		}
	}
	if (argc == optind) {
		(void) fprintf(stderr, "Usage: %s algorithm [file...]\n", *argv);
		return EXIT_FAILURE;
	}
	if ((alg = find_algorithm(argv[optind])) == NULL) {
		errx(EXIT_FAILURE, "No such algorithm `%s'", argv[optind]);
	}
	rval = EXIT_SUCCESS;
	if (argc == optind + 1) {
		if (!(*alg->func)(NULL)) {
			warn("stdin");
			rval = EXIT_FAILURE;
		}
	} else {
		for (i = optind + 1 ; i < argc ; i++) {
			if (!(*alg->func)(argv[i])) {
				warn("%s", argv[i]);
				rval = EXIT_FAILURE;
			}
		}
	}
	return rval;
}
